OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs

```diff
@@ -11,55 +11,150 @@
 
-		readonly IEnumerable<SupportPowerInstance> powers;
-		Pair<string, Color>[] texts;
+		readonly Dictionary<Player, SupportPowerManager> spManagers;
+		readonly SortedDictionary<int, Player[]> playersByTeam;
 
-		[ObjectCreator.UseCtor]
-		public SupportPowerTimerWidget(World world)
+		readonly string fontName = "Bold";
+		readonly SpriteFont font;
+		readonly float yIncrement, powerPos, timePos;
+
+		// Represents one line of text
+		struct Text
 		{
-			powers = world.ActorsWithTrait<SupportPowerManager>()
-				.Where(p => !p.Actor.IsDead && !p.Actor.Owner.NonCombatant)
-				.SelectMany(s => s.Trait.Powers.Values)
-				.Where(p => p.Instances.Any() && p.Info.DisplayTimer && !p.Disabled);
+			public Color PlayerColor;
+			public string Player;
+			public string Power;
+			public string RemainingTime;
+			public bool Ready;
+			public string Team;		// Used only for team label
 		}
 
-		public override void Tick()
+		readonly List<Text> display;
+		readonly bool noPowers;
+		bool flashing, needTeamLabel, showPlayer;
+		Color timerColor;
+
+		readonly World world;
+
+		[ObjectCreator.UseCtor]
+		public SupportPowerTimerWidget(World world)
 		{
-			texts = powers.Select(p =>
-			{
-				var time = WidgetUtils.FormatTime(p.RemainingTime, false);
-				var text = Format.F(p.Info.Description, time);
-				var color = !p.Ready || Game.LocalTick % 50 < 25 ? p.Instances[0].Self.Owner.Color.RGB : Color.White;
-				return Pair.New(text, color);
-			}).ToArray();
+			spManagers = new Dictionary<Player, SupportPowerManager>();
+			foreach (var tp in world.ActorsWithTrait<SupportPowerManager>()
+				.Where(tp => !tp.Actor.IsDead && !tp.Actor.Owner.NonCombatant))
+				spManagers[tp.Actor.Owner] = tp.Trait;
+
+			var availablePowers = world.Map.Rules.Actors.Values
+				.SelectMany(ai => ai.Traits.WithInterface<SupportPowerInfo>())
+				.Where(i => i.DisplayTimer).ToArray();
+
+			noPowers = !spManagers.Keys.Any() || !availablePowers.Any();
+			if (noPowers)
+				return;
+
+			playersByTeam = new SortedDictionary<int, Player[]>();
+			foreach (var team in spManagers.Keys.GroupBy(p => world.LobbyInfo.ClientWithIndex(p.ClientIndex).Team))
+				playersByTeam[team.Key] = team.ToArray();
+
+			font = Game.Renderer.Fonts[fontName];
+			yIncrement = font.Measure(" ").Y + 5;
+			var padding = font.Measure("  ").X;
+			var maxPlayerWidth = spManagers.Keys.Max(p => font.Measure(p.PlayerName).X);
+			var maxPowerWidth = availablePowers.Max(i => font.Measure(i.Description).X);
+			powerPos = maxPlayerWidth + padding;
+			timePos = powerPos + maxPowerWidth + padding;
+
+			display = new List<Text>();
+			this.world = world;
 		}
 
 		public override void Draw()
 		{
-			if (!IsVisible() || texts == null)
+			if (!IsVisible() || noPowers)
 				return;
 
-			var y = 0;
-			foreach (var t in texts)
+			// Generate texts once in every 10 WorldTicks,
+			// which is 0.4 seconds in game time.
+			if (world.WorldTick % 10 == 0)
```

Comment:
>it must be a race condition when Tick is faster than rendering

Between your draws, the Ticks already pass. Either that, either the UI is separated from the world in this case.

Considering that the ready-text-blinking code uses Game.LocalTick instead, I'd assume that's much reliable than this. As I said it already.